### DISCUSSION TITLE
MustardUI 0.30

### DISCUSSION THREAD

**Discussion by Mustard2** [Mustard2/MustardUI#120] [Ideas]
Hi!

A new version of Blender is coming (4.0), and therefore also a new version of MustardUI is almost ready (0.30), see #119 .

One of the biggest new feature is that MustardUI will now be an add-on only, This means that both creators should now install MustardUI as an add-on. But why this was necessary? There are some reasons:

- **Code Maintenance**; as the script was reaching 10k, the maintenance of the whole code was becoming a nightmare! Every time I had to change something I had to remember where I put the functions/operators/definitions in the code, and the IDE was not helping a lot ue to the huge amount of things on the same file. Separating the script in several files was the only way, and unfortunately this means that MustardUI can not be used as a script anymore (only single file script can run in Blender).

![image](https://github.com/Mustard2/MustardUI/assets/60780948/d6190c88-5470-4466-b964-9b3f7fe53c67)

- **Community engagement**: a lot of people asked me to contribute to the project, but I must admit it was a nightmare to search things in the 10k lines file. Now it should be easier to find the various parts of the UI, and I hope this will make you try and contribute directly on the code. This is the structure you can find in the new version.

![image](https://github.com/Mustard2/MustardUI/assets/60780948/ef329fd5-9b4a-4ee4-94a1-29a73b418aae)

- **Script/Add-on confusion**: Using MustardUI was always confusing, especially for new people. I constantly get PMs asking me how to install MustardUI from users of the model. Now MustardUI will work only as an add-on, hopefully removing any confusion.
- **Security**: No registration of any script is required now, and this is a major security step forward! In fact, I had no control over the script once it was downloaded from the repository. Somebody could have maliciously changed it before shipping their models, potentially causing harm to the system were the model was used. With the addon, people will be forced to install it from a known and trusted source (GitHub) directly, avoiding the possibility of model creators to include malicious scripts in their files. In fact, install add-ons only from trusted sources (in general, directly from the creator website). *I am aware of someone using the script shipping malwares, luckily they weren't so smart and a Windows Defender could find and block it.*
- **Performance**: In general add-ons should perform better, and avoid issues where several scripts of MustardUI could be available in the same scene.

The major drawback about this that made me wait a lot to rewrite this as an add-on is the fact that now users should install MustardUI. Installing add-ons in Blender is not always straightforward, but I will provide a tutorial for the installation so model creators can share that directly with their users. :)

Please let me know if you have any feedback/idea about this!

For those interested, coding add-ons in Blender is extremely easy with PyCharm (thanks Jetbrains for the free Community Edition!).
Some useful resources:

- https://github.com/BlackStartx/PyCharm-Blender-Plugin
- https://b3d.interplanety.org/en/using-external-ide-pycharm-for-writing-blender-scripts/

**Comment by Ghost76150** [comment]
Hi, 
I can't activate this add on MustardUi, i'm having this issue if i try :
![issue](https://github.com/Mustard2/MustardUI/assets/130587553/50e77958-7a3f-44bf-80cb-12eddc625b32)

Best Regards.

**Reply by Mustard2** [reply]
Hi! Are you using Blender 4.0.2? Plus, I see two MustardUI.
I would suggest to remove both of them, restart Blender, and install the latest version.

**Comment by Ghost76150** [comment]
Hi, i was using the previous version of Blender 3.4 and also see that Mustard was in the list twice

i guess under c:\Program Files\Blender Foundation\...

and also under my C:\Users\Utilisateur\AppData\Roaming\Blender Foundation\Blender\3.4\scripts\MustardUI

in which directory ? should it be stored  : **C:\Program Files\...** or under **Roaming\...** etc ?

**Reply by Mustard2** [reply]
I think the correct one should be the second one, but to be sure I would remove both, restart Blender and then install it again from Blender only once.

**Reply by ilego1** [reply]
Apologies for messaging you here I messaged u in discord thanking you for the UI and asking a question.

**Comment by Mustard2** [comment]
> Apologies for messaging you here I messaged u in discord thanking you for the UI and asking a question.

@ilego1 Thank you! If you have any question on the UI, it's easier here. Even if it's not an issue, feel free to open a discussion at https://github.com/Mustard2/MustardUI/issues

**Reply by ilego1** [reply]
Thanks my apologies for late reply I will say though that I do like your UI
